Initial version

iBeizsley opened the pull request.

Commit:
Initial project

Commit:
Implemented WorkingDayService

Commit:
Implemented WorkingDayServiceBuilder

Commit:
Implemented DayOfTheWeekWorkingDaySource

Commit:
Implementet HTTPWorkingDaySource & FileWorkingDaySource

Commit:
Implemented GovUK Bank Holiday JSON source

Commit:
Added working day addition and subtraction methods to WorkingDayService

Commit:
Improved UK Gov Bank Holiday JSON source logic

Commit:
Fixed multiple-source IsWorkingDay logic

Commit:
Added tests for WorkingDayService

Commit:
Added tests for StringSource & UK Gov Bank Holiday Json Parsing

Commit:
Warning fix

lparkermg committed:
Updated Microsoft Test SDK to 15.7.2

oscarhirst committed:
Improved the documentation in WorkingDayServiceBuilder.cs

oscarhirst committed:
Improved the documentation of IWorkingDayService.cs

oscarhirst committed:
Improved documentation of WorkingDayService.cs

oscarhirst committed:
Improved documentation of working day extension method classes:
WorkingDaySourceExtensions.cs
DayOfTheWeekSourceExtensions.cs

oscarhirst committed:
Improved documentation of DayOfTheWeekWorkingDaySource.cs

oscarhirst committed:
mproved documentation of FileSourceExtensions.cs

oscarhirst committed:
Improved documentation of:
DayOfTheWeekSourceExtensions.cs
FileSourceExtensions.cs
FileWorkingDaySource.cs
StringSourceExtensions.cs
StringWorkingDaySource.cs

oscarhirst committed:
Improved documentation of HTTP and JSON source classes and methods

jbradford commented on src/CR.WorkingDayService/WorkingDayServiceBuilder.cs:
```diff
@@ -0,0 +1,96 @@
+        /// </summary>
+        /// <param name="source">The <see cref="IWorkingDaySource"/> to use.</param>
+        /// <returns>The same instance of the <see cref="WorkingDayServiceBuilder"/> using only the <see cref="IWorkingDaySource"/> which was passed in. All all of its current sources will be removed.</returns>
+        public WorkingDayServiceBuilder UseSource(IWorkingDaySource source)
```
I don't see the point of `UseSource` vs `AddSource`, `WithSource` would be much simpler to use.
For a single source `WorkingDayService` then you just have a single `WithSource`, then for a multi source `WorkingDayService` you would chain `WithSource` calls.
With `UseSource` and `AddSource` you could create some very weird looking calls.

```
WorkingDayServiceBuilder.New().AddSource(a).AddSource(b).UseSource(c).Build();
```
The `WorkingDayService` would end up only having a source of `c`, but without looking at the comments it would be pretty difficult to understand why.

jbradford commented on src/CR.WorkingDayService/WorkingDayService.cs:
```diff
@@ -0,0 +1,102 @@
+        {
+            for (var i = 0; i < numberToAdd; i++)
+            {
+                do
```
This is the same logic in `NextWorkingDay` and `PreviousWorkingDay`, we could implement `AddWorkingDays` and `SubtractWorkingDays` by calling the Next and Previous methods the correct number of times.

jbradford commented on src/CR.WorkingDayService/IWorkingDaySource.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// A common interface for implementing a Source of information on days' working/non-working status.
+    /// </summary>
+    public interface IWorkingDaySource
```
Is there a benefit to having this as an interface over a base class? With the base class we could have the implementation of `IsNonWorkingDay` implemented directly rather than as an extension method.

jbradford reviewed (commented):
Checked the base classes and left some comments (for discussion perhaps), haven't looked at the implementations yet.

oscarhirst committed:
Simplified WorkingDayService's Add/Subtract by calling Next/Prev

oscarhirst committed:
IWorkingDaySource + Extensions => WorkingDaySource.cs

oscarhirst committed:
Corrected a FileWorkingDaySource.cs inherit doc ref

oscarhirst committed:
Partially fixed bug - requires more work

oscarhirst committed:
Corrected StringNWDSource's use of CheckAction

oscarhirst committed:
Updated the AddWorkingDayTests

oscarhirst committed:
Updated some WorkingDayServiceTests:
IsNonWorkingDayTests
SubtractWorkingDayTests

oscarhirst committed:
Updated tests

oscarhirst committed:
Corrected Service's Add/Sub. methods to behave in accordance with OM

oscarhirst committed:
Corrected the CheckAction description

jbradford commented on src/CR.WorkingDayService/WorkingDayService.cs:
```diff
@@ -0,0 +1,121 @@
+        /// </summary>
+        /// <param name="sources">The <see cref="NonWorkingDaySource"/>s to use to determine whether a given <see cref="DateTime"/> is on a working day, or a non-working day.
+        /// If no sources are specified, all days are considered working days.</param>
+        public WorkingDayService(IReadOnlyCollection<NonWorkingDaySource> sources) => _sources = sources;
```
Surely a `WorkingDayService` with no sources is actually an error? If you want one with every day as a working day you should explicitly have to create that.

jbradford commented on src/CR.WorkingDayService/WorkingDayServiceBuilder.cs:
```diff
@@ -0,0 +1,96 @@
+        /// </summary>
+        /// <param name="source">The <see cref="NonWorkingDaySource"/> to use.</param>
+        /// <returns>The same instance of the <see cref="WorkingDayServiceBuilder"/> using only the <see cref="NonWorkingDaySource"/> which was passed in. All all of its current sources will be removed.</returns>
+        public WorkingDayServiceBuilder UseSource(NonWorkingDaySource source)
```
I still feel this is wrong, there is no need for a `UseSource` and `AddSource`, we should have a single `WithSource` that looks like `AddSource`, see my previous comment on the last review about this for reasoning.

jbradford replied:
This should apply to anywhere we've got separate `Use` and `Add` sources.

jbradford commented on src/CR.WorkingDayService.DayOfWeekSource/WeekendDaysNonWorkingDaySource.cs:
```diff
@@ -0,0 +1,37 @@
+    /// <summary>
+    /// An implementation of <see cref="NonWorkingDaySource"/> which uses the <see cref="DayOfWeek"/> of the provided <see cref="DateTime"/> to determine whether the day is a Working Day or Non-Working Day.
+    /// </summary>
+    public sealed class WeekendDaysNonWorkingDaySource : NonWorkingDaySource
```
We should have a more generic `DayOfWeekNonWorkingDaySource` then in the Extensions can provide shortcuts for initialising this with weekend days, this `Source` is currently inflexible.

jbradford commented on src/CR.WorkingDayService.StringSource/StringNonWorkingDaySource.cs:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        /// <summary>
+        /// Gets the action used to check whether a <see cref="DateTime"/> is on a Working Day (using the <see cref="StringNonWorkingDaySource{T}"/>).
```
Doesn't this now check it ISN'T on a working day?

jbradford commented on src/CR.WorkingDayService.GovUkJsonSource/GovUkBankHolidayJsonSourceExtensions.cs:
```diff
@@ -0,0 +1,60 @@
+        private static NonWorkingDaySource GovUkBankHolidayJsonSource(TimeSpan refreshTime) => new HttpNonWorkingDaySource<List<DateTime>>(
+            new HttpRequestMessage(HttpMethod.Get, "https://www.gov.uk/bank-holidays.json"),
+            GovUkBankHolidayJsonParse,
+            (dateTime, list) => !list.Contains(dateTime.Date),
```
I *think* this is now wrong, although with all the abstractions it's hard to tell.

jbradford reviewed (dismissed):
Not a full review. I think all the code needs a bit of a review in light of the switch to `NonWorkingDay` and it's quite hard to follow the naming of some things, and some comments don't seem to match what they're actually doing.

rcknight commented on src/CR.WorkingDayService/WorkingDayServiceBuilder.cs:
```diff
@@ -0,0 +1,96 @@
+        /// Builds a <see cref="WorkingDayService"/> with the configured set of <see cref="NonWorkingDaySource"/>s.
+        /// </summary>
+        /// <returns>A <see cref="WorkingDayService"/> which uses the <see cref="NonWorkingDaySource"/>s in the <see cref="WorkingDayServiceBuilder"/>'s <c>_sources</c>.</returns>
+        public WorkingDayService Build() => new WorkingDayService(_sources.ToList());
```
You can make a builder that doesn't require you to call .Build() on the end with an implicit conversion operator ... something like:

```
public static implicit operator WorkingDayService(WorkingDayServiceBuilder builder)
{
    return new WorkingDayService(builder.sources.ToList());
}
```

oscarhirst committed:
Removed UseSource method, renamed AddSource to WithSource

jbradford commented:
Why have we sealed all the classes? Is there a reason why we wouldn't want external users (or even ourselves) being able to extend the FileSource (for example)?

oscarhirst committed:
Added generic DaysOfTheWeekSource
corrected descriptions of 'if it is a working day' => 'non-working day'
Fixed the Json source to source conversion
removed builder.build() added implicit converter
Prevented building a service with no sources

oscarhirst committed:
Unsealed classes

Commit:
Update package info

Commit:
Final cleanup & behaviour tweaks

lparkermg reviewed (dismissed):
LGTM

Commit:
'Strinc'

Commit:
Proofread & fixed typos & inconsitencies

Commit:
Set up for building & nuget packaging

Commit:
Package version range

Commit:
Fix package reference versions for Nuget.org

Commit:
Finalize dependency versions

Commit:
CR.WorkingDayService -> CR.Tools.WorkingDayService

Commit:
Update build scripts for new csproj path

Commit:
Updated namespacing

Commit:
Fix Readme markdown

jbradford committed:
Update Logo URL

rcknight committed:
Merge pull request #2 from cognisant/building

Setup for building Nuget package
